src/smtml/expr.ml

```diff
@@ -320,13 +303,66 @@ let exists vars body = binder Exists vars body
 
 let raw_unop ty op hte = make (Unop (ty, op, hte)) [@@inline]
 
+let is_val_real e =
+  match view e with
+  | Val (Real x) -> Float.is_finite x
+  | _ -> false
+
+let normalize_eq_or_ne op (ty', e1, e2) =
+  let is_real = is_val_real e1 && is_val_real e2 in
+  let make_relop lhs rhs = Relop (ty', op, lhs, rhs) in
+  let ty,ty2 = ty e1,ty e2 in
+  if Ty.equal ty ty2 then
```

Comment:
```suggestion
  assert (Ty.equal ty ty2);
```
At some point in the future, I plan to ensure type safety during expression construction. So, I expect this invariant to always hold, so we can replace the `if` with an `assert` as a sanity check.